src/components/graph-sonification.tsx

```diff
@@ -22,153 +52,284 @@ export const GraphSonification = ({ graphToSonify, onResetGraphToSonify }: IProp
+        restartTransport();
+        frameIdRef.current = requestAnimationFrame(step);
+      }
+    };
+
+    frameIdRef.current = requestAnimationFrame(step);
+  }, [durationRef, selectedGraph, handlePlayEnd, scheduleTones]);
+
+  const prepareSonification = useCallback(async () => {
+      if (!selectedGraph) return;
+
+      if (currentGraphIdRef.current && currentGraphIdRef.current !== selectedGraph.id) {
+        await removeRoiAdornment(currentGraphIdRef.current);
       }
 
-      currentGraphIdRef.current = graphDetails.id;
+      currentGraphIdRef.current = selectedGraph.id;
+      await Tone.start();
+      scheduleTones();
 
       await codapInterface.sendRequest({
         action: "create",
-        resource: `component[${graphDetails.id}].adornment`,
+        resource: `component[${selectedGraph.id}].adornment`,
         values: {
           type: "Region of Interest",
-          primary: { "position": "0%", "extent": 0.05 },
+          primary: { "position": "0%", "extent": "0.05%" }, // 0.05% consistently approximates 1 pixel
           secondary: { "position": "0%", "extent": "100%" }
         }
       });
 
-      const pitchAttr = graphDetails.yAttributeName;
-      const timeAttr = graphDetails.xAttributeName;
-
-      const allItemsRes = await getAllItems(graphDetails.dataContext);
-      const allItems = allItemsRes.values;
+      restartTransport();
+      animateSonification();
+  }, [animateSonification, selectedGraph, scheduleTones]);
 
-      // Asign pitch and time values to the dataset items
-      // Do not include items that are missing values for either attributes
-      const validItems = pitchAttr && timeAttr
-        ? allItems.filter((item: CodapItem) => item.values[pitchAttr] !== "" && item.values[timeAttr] !== "")
-        : allItems.filter((item: CodapItem) => item.values[timeAttr] !== "");
+  const mapValuesToTimeAndPitch = async (graphDetails: Record<string, any>) => {
+    const pitchAttr = graphDetails.yAttributeName;
+    const timeAttr = graphDetails.xAttributeName;
+    const allItemsRes = await getAllItems(graphDetails.dataContext);
+    const allItems = allItemsRes.values;
 
-      const pitchValues = validItems.map((item: CodapItem) => item.values[pitchAttr]);
-      const timeValues = validItems.map((item: CodapItem) => item.values[timeAttr]);
+    // Do not include items that are missing values for the attributes.
+    const validItems = pitchAttr && timeAttr
+      ? allItems.filter((item: CodapItem) => item.values[pitchAttr] !== "" && item.values[timeAttr] !== "")
+      : allItems.filter((item: CodapItem) => item.values[timeAttr] !== "");
+    const pitchValues: number[] = validItems.map((item: CodapItem) => item.values[pitchAttr]);
+    const timeVals: number[] = validItems.map((item: CodapItem) => item.values[timeAttr]);
+    const timeRange = graphDetails.xUpperBound - graphDetails.xLowerBound || 1;
+    const pitchRange = graphDetails.yUpperBound - graphDetails.yLowerBound || 1;
```

Comment:
The time and pitch scale based on the min and max of the data, but I think it should be based on the min and max of the visible area of the graph.  If you make a graph and then drag it left or right so that only some of the points are visible, the sound no longer matches with what the ROI bar is tracing.

Also If you scale the graph so that it covers a very narrow range in the display, the pitches still cover the full range.

Reply:
The upperBound/lowerBound properties actually do change when you drag the axes. Now that the state is more reactive the sonification should match what the bar is tracing.

Reply:
Yes, this works nicely now -- subject to a 5-second polling delay for the update to be noticed.